tests/bwc/test_recovery.py

```diff
@@ -0,0 +1,491 @@
+import time
+import unittest
+
+from crate.client import connect
+import random
+from random import sample
+
+from crate.qa.tests import NodeProvider, insert_data
+from bwc.test_rolling_upgrade import UpgradePath
+
+
+class RecoveryTest(NodeProvider, unittest.TestCase):
+    """
+    In depth testing of the recovery mechanism during a rolling restart.
+    Based on org.elasticsearch.upgrades.RecoveryIT.java
+    """
+
+    def test(self):
+        self._run_tests(
+            [
+                UpgradePath('4.2.x', '4.3.x'),
+                UpgradePath('4.3.x', 'latest-nightly'),
+            ],
+            [
+                self._test_relocation_with_concurrent_indexing,
+                self._test_recovery,
+                self._test_update_docs,
+                self._test_recovery_closed_index,
+                self._test_closed_index_during_rolling_upgrade,
+                self._test_relocation_with_concurrent_indexing,
+            ]
+        )
```

Comment:
I'm not sure I like this pattern of having a single test entry point to manually dispatch the test cases.

This breaks unit test runner functionality to run individual test cases.

Reply:
The purpose behind that was to have the test running with multiple upgrade paths, similar to `test_rolling_upgrades`. However i will find a better solution that this.

Reply:
It looks nice, but also seems to break tooling (https://github.com/wolever/parameterized/issues/37, https://github.com/Microsoft/PTVS/issues/1669)

Not sure what's worse. Hand-rolling the upgrade path within each test case or breaking IDE/editor functions to launch the individual test cases?

Reply:
Good point, i simplified it now and removed the external dependency.